File: python/2841.py
class Solution(object):
    def maxSum(self, nums, m, k):
        """
        :type nums: List[int]
        :type m: int
        :type k: int
        :rtype: int
        """
        num_map = {} # 当前窗口各元素出现的次数
        cnt = 0 # 当前窗口数字总和
        """计算第一个窗口的不系统元素个数和窗口数字总和"""
        for i in range(0, k):
            if nums[i] not in num_map:
                num_map[nums[i]] = 1
            else:
                num_map[nums[i]] += 1
            cnt += nums[i]
        ans = cnt if len(num_map) >= m else 0 # 结果
        """移动窗口：cnt变化、字段变化、cur变化、结果变化"""
        for i in range(k, len(nums)):
            cnt -= nums[i - k]
            if num_map[nums[i - k]] == 1:
                num_map.pop(nums[i - k])
            else:
                num_map[nums[i - k]] -= 1
            if nums[i] not in num_map:
                num_map[nums[i]] = 1
            else:
                num_map[nums[i]] += 1
            cnt += nums[i]
            if len(num_map) >= m and cnt > ans:
                ans = cnt
        return ans
            

File: cpp/2841.cpp
#include <iostream>
#include <vector>
#include <map>
using namespace std;

class Solution {
public:
    long long maxSum(vector<int>& nums, int m, int k) {
        long long cur = 0; // 当前窗口总和
        long long ans = 0; // 最优解
        map<int, int> cnt; // 字典，记录窗口内每个元素出现的次数
        int len = nums.size();
        for(int i = 0; i < len; ++i) {
            // 窗口增加右端点
            cur += nums[i];
            ++cnt[nums[i]];
            // 确认窗口左端点的位置
            int left = i - k + 1;
            if (left < 0) { // 不是完整的窗口、继续延长窗口
                continue;
            }
            // 更新答案
            if (cnt.size() >= m && cur > ans) {
                ans = cur;
            }
            // 窗口离开左端点
            cur -= nums[left];
            if(--cnt[nums[left]] == 0) {
                cnt.erase(nums[left]);
            }
        }
        return ans;
    }
};